### DISCUSSION TITLE
Custom exception hook doesn't get called in Airflow tasks

### DISCUSSION THREAD

**Discussion by instanceofmel** [apache/airflow#35797] [Q&A]
Hello,

we're using Airflow for data serialization and inside a DAG we import our ETL library. This ETL library has logic for querying some data from our DB, doing some transformations and serializing in a specific format. During those steps, exceptions can occur and we want to know which data object was being processed when the exception occured. That's why our library 'installs' a custom exception hook. We have overwritten `sys.excepthook` with something like:

```python
# etl-lib/__init__.py

current_data_id : int = None

default_excepthook = sys.excepthook

def set_current_context(data_id):
  global current_data_id
  current_data_id = data_id

def exception_hook(*args, **kwargs):
  default_excepthook(*args, **kwargs)
  sys.stderr.write(f"data_id = {current_data_id}")

sys.excepthook = exception_hook
```

Now every time some processing is being done on the data, we call `set_current_context(data.id)` beforehand. That works fine when we run the ETL library in our CLI. We can see the faulty `data_id` when an exception occurs.

However, when we use the library inside our DAG (we use taskflow and use the library inside the tasks) and an exception occurs, the `data_id` doesn't get printed in the airflow logs (= the custom exception hook doesn't get called)

I guess it's because while the exception bubbles up, it's catched somewhere from Airflow (`PythonOperator`?) and then the stacktrace is printed from another context (and here the excepthook is not installed). Does anyone know how to bypass this?

Help would be greatly appreciated 🙏

**Comment by potiuk** [comment]
You would need to install excepthook in each task. Not sure where you are installing the hook, but generally when the task gets executed, The DAG is parsed in a separately forked process in the worker (or POD) and only the "execute()" method of that task is called. And - depending on the way how forking/spawning and some optimizations (configurable via Airflow configuration (https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#mp-start-method , https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#parsing-pre-import-modules among others) some code is executed in parent process (for example airflow imports) , and some other code in the forked subprocess ) 

What finally gets executed inside the forked process are:

* parsing the DAG top level code (but some imports might come already pre-imported before the fork happened in the parent process) 
* running the "execute" method of the particular task (and only that task) that gets executed (which in your case is the method decorated with `@task`)

We also do weird things with stderr like replacing stderr with in-memory redirection to task logger and this might mess a bit with you writing directly to stderr that might go through the forking I explained above.

I am not sure what happens with excepthook when fork happens, but it might be that it is reset when fork happens. Also it might be that except hook installed in the parent prints stuff to stderrthat was opened in the parent proces (and is a different handle in the forked one) - so a lot of things **could** go wrong.

I think what you can try is to:

1) change the place where you install excepthook - to be inside task (if only for testing it might let you debug what exactly happens
2) maybe you can simply use the logger (from task can  use "airflow.task" logger to log to the task log. I think if you the logger it might properly log things to the right file regardless of where excepthook was installed.

That would be the places I'd look. Also you can debug it in the way that you can open a file and log to it every time you get such exception ("/tmp/test_log.txt" say). This is the way I usually test if the things are even executed when I am not sure where my stderr is going to.
